### PR TITLE
Dev-4019 New Dates Section & Award Status Section

### PR DESCRIPTION
**Pull request by jonhill13**
**High level description:**

Updates the Award Page to include the new dates section from the new mock. Adds the Award Status section to the Award title.

**Technical details:**

> * refactor Two Rectangles to accept any pattern
> 
> * update idv activity chart to use new patters component
> 
> * update award summary helper to include days
> 
> * update progress bar to inlcude three patterns for contracts
> 
> * update css
> 
> * update progress bar to include triangle
> 
> * update progression text
> 
> * update progress bar to include vertical today line
> 
> * create award status
> 
> * add one day to award status
> 
> * update tests
> 
> * update css

**JIRA Ticket:**
[DEV-4019](https://federal-spending-transparency.atlassian.net/browse/DEV-4019)

**Mockup:**
https://bahdigital.invisionapp.com/share/H4IAC8SREXU#/screens

The following are ALL required for the PR to be merged:

Author:
- [x] Linked to this PR in JIRA ticket
- [x] Scheduled demo including Design/Testing/Front-end OR Provided instructions for testing in JIRA and PR `if applicable`
- [x] Verified cross-browser compatibility
- [x] Verified mobile/tablet/desktop/monitor responsiveness
- [x] Verified that this PR does not create any *new* accessibility issues (via Axe Chrome extension)
- [x] Added Unit Tests for methods in Container Components, reducers, and helper functions `134 -> 133 in dev with award award/CONT_AWD_HC101318FB025_9700_HC101317A0004_9700`
- [x] All componentWillReceiveProps, componentWillMount, and componentWillUpdate in relevant child/parent components/containers replaced with [future compatible life-cycle methods](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html) per [JIRA item 3339](https://federal-spending-transparency.atlassian.net/browse/DEV-3339)
- [N/A] [API contract](https://github.com/fedspendingtransparency/usaspending-api/tree/dev/usaspending_api/api_contracts) updated `if applicable`

Reviewer(s):
- [x] Design review complete `if applicable`
- [N/A] [API #1234](https://github.com/fedspendingtransparency/usaspending-api/pull/1234) merged concurrently `if applicable`
- [ ] Code review complete

### CODE REVIEW

File: src/js/components/award/shared/AwardStatus.jsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+const AwardStatus = ({ dates, awardType }) => {
+    const { startDate, endDate, currentEndDate } = datesByAwardType(dates, awardType);
+    const badDates = isBadDates({ startDate, endDate, currentEndDate }, awardType);
+    const contract = isContract(awardType);
+    const awardStatus = () => {
+        if (badDates) return '';
+        const today = moment();
+        let end = endDate;
+        if (contract) end = currentEndDate;
+        end = end.add(1, 'd');
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/AwardStatus.jsx]: comment here por favor as to why we add a day

File: src/js/components/award/shared/AwardStatus.jsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        // Completed or Closed
+        if (today.isAfter(end)) status = (awardType === 'idv') ? 'Closed' : 'Completed';
+        return status;
+    };
+
+    const timeRemaining = () => {
+        if (badDates) return '';
+        let dateToCompare = contract ? currentEndDate : endDate;
+        if (!dateToCompare) return '';
+        dateToCompare = dateToCompare.add(1, 'd');
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/AwardStatus.jsx]: Ditto

File: src/js/components/award/shared/overview/AwardDates.jsx [hunk 3]
```diff
@@ -3,152 +3,164 @@
...
-        return [
+        return compact([
             titles[awardType][0],
             titles[awardType][1],
             titles[awardType][2]
-        ];
+        ]);
+    }
+
+    progressDescriptions = () => {
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/AwardDates.jsx]: Lets move all these variables outside the class.

File: src/js/components/award/shared/overview/ProgressBar.jsx [hunk 4]
```diff
@@ -0,0 +1,560 @@
+/**
+ * ProgressBar.jsx
+ * Created by Jonathan Hill 11/26/19
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/ProgressBar.jsx]: Don't be lyin.

File: src/js/components/award/shared/overview/ProgressBar.jsx [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+        this.setTextDiv = (element) => {
+            this.textDiv = element;
+            this.positionText();
+        };
+    }
+
+    // recreate the progress bar on window resize
+    componentDidUpdate(prevProps) {
+        if (prevProps.width !== this.props.width) {
+            this.start();
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/ProgressBar.jsx]: Why not just remove `this.start()` and invoke `this.validateDomainAndMilestones`?

File: src/js/components/award/shared/overview/ProgressBar.jsx [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+            currentProgress,
+            domain
+        } = this.props;
+        // validation of domain
+        let badDomainData = this.props.badDomainData || false;
+        pull(domain, null, undefined, '');
+        if (domain.length !== 2) {
+            badDomainData = true;
+        }
+        if (domain[0] > domain[1]) badDomainData = true;
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/ProgressBar.jsx]: might be helpful to use array destructiong here and give domain[0] a nice var name.

```javascript
const [startDate, endDate] = pull(domain, null, undefined, ''); 
```

File: src/js/components/award/shared/overview/ProgressBar.jsx [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+        // handle bad dates
+        this.props.milestones.forEach((milestone) => {
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/ProgressBar.jsx]: lets create a new var here:


```javascript
const parsedMilestones = this.props.milestones.map()
```

`forEach` is overrated anyway and mutating arrays is too.

File: src/js/components/award/shared/overview/ProgressBar.jsx [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+            if (milestone.data < domain[0]) {
+                badDomainData = true;
+            }
+        });
+        // validation of current progress
+        if (!currentProgress) badDomainData = true;
+        this.setState({
+            currentProgress,
+            badDomainData
+        }, this.createXScales);
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/ProgressBar.jsx]: FWIW I object to the use of `let` as it mutates references. Instead, I'd recommend executing a return under any condition when you would reassign the value of a `let` variable.

☝️ in conformity to principle of immutability.

File: src/js/components/award/shared/overview/ProgressBar.jsx [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+            x1: position,
+            x2: position,
+            y1: 0,
+            y2: heightOfProgressBar,
+            visualizationStart: position,
+            visualizationEnd: position,
+            display: true
+        };
+        // show progress vertical line
+        milestoneData.forEach((milestone, index, array) => {
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/ProgressBar.jsx]: If mutating an array, don't use `.forEach`. Instead, create a new array and use `.map`

File: src/js/components/award/shared/overview/ProgressBar.jsx [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+        // verify currently in progress
+        if (!(domain[0] < currentProgress && currentProgress < domain[1])) {
+            progressVerticalLineData.display = false;
+        }
+        this.setState({ progressVerticalLineData }, this.progressTriangleData);
+    }
+    // progress triangle
+    // the points are three coordinate pairs in the form of x,y
+    // they are in order from left to right as a string separated by a space
+    progressTriangleData = () => {
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/ProgressBar.jsx]: nice comments

File: src/js/components/award/shared/overview/ProgressBar.jsx [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+            // Therefore, we need to complete the following steps to know if the text overflows:
+            // 1. Get the total pixels between the current progress and then end of the progress
+            // between the cirlces.
+            // 2. Get the total pixels from the end of the progress between the circles and the
+            // end of the SVG.
+            // 3. Determine if half the text width (since we position the text directly in the
+            // center of progress) is greater than the total pixels left in the SVG.
+            // 4. If #3 is true, we must then determine the amount of pixel overflow by subtracting
+            // half the text width and the pixels available in the SVG and update the new postion of
+            // the text.
```
Comment (MaxwellKendall) [inline src/js/components/award/shared/overview/ProgressBar.jsx]: in this case, use /**/ comment syntax

File: src/js/helpers/awardSummaryHelper.js [hunk 12]
```diff
@@ -56,3 +62,34 @@ export const getAwardTypeByRecordtypeCountyAndState = (
     // IDV or contract
     return 'nonFinancialAssistance';
 };
+
+export const datesByAwardType = (dates, awardType) => {
```
Comment (MaxwellKendall) [inline src/js/helpers/awardSummaryHelper.js]: The naming convention `xByZ` implies a map that is keyed by a certain thing and has the value of a certain thing.

so `datesByAwardType` should look like this:

```javascript
{ contract: dates, asst: dates, idv: dates }
```

Would rename this map to conform to this convention or just use an array or possibly something like: `datesByDateType`

File: src/js/components/award/shared/overview/ProgressBar.jsx (hunk 4 above)
Comment (jonhill13) [reply src/js/components/award/shared/overview/ProgressBar.jsx]: lol, that is when i started it, with the first iteration of the new design of the dates section

File: src/js/components/award/shared/overview/ProgressBar.jsx (hunk 9 above)
Comment (jonhill13) [reply src/js/components/award/shared/overview/ProgressBar.jsx]: we are not mutating an array here. We are just updating the `progressVerticalLineDataObject` `display` property. We need to loop through the milestones.

**Pull request merged**